src/compiler/extensions.ts

```diff
@@ -0,0 +1,379 @@
+            const currentDirectory = host.getCurrentDirectory ? host.getCurrentDirectory() : "";
+            const extMap: Map<string> = {};
+            forEach(extensionNames, name => {
+                const resolved = resolveModuleName(name, combinePaths(currentDirectory, "tsconfig.json"), options, host, /*loadJs*/true).resolvedModule;
+                if (resolved) {
+                    extMap[name] = resolved.resolvedFileName;
+                }
+            });
+            return extMap;
+        }
+
+        function collectCompilerExtensions(): ExtensionCollectionMap {
+            const profileLevel = options.profileExtensions;
+            const extensionLoadResults = map(extensionNames, (name) => {
+                const resolved = resolvedExtensionNames[name];
+                let result: any;
+                let error: any;
+                if (!resolved) {
+                    error = new Error(`Host could not locate extension '${name}'.`);
+                }
+                if (resolved && host.loadExtension) {
+                    try {
+                        if (profileLevel) {
+                            startProfile(name, name);
+                            if (profileLevel >= ProfileLevel.Full) trace(Diagnostics.PROFILE_Colon_Extension_0_begin_1, name, "load");
+                        }
+                        result = host.loadExtension(resolved);
+                        if (profileLevel) {
+                            completeProfile(name);
+                            if (profileLevel >= ProfileLevel.Full) trace(Diagnostics.PROFILE_Colon_Extension_0_end_1_2_ms, name, "load", perfTraces[name].length.toPrecision(5));
+                        }
+                    }
+                    catch (e) {
+                        error = e;
+                    }
+                }
+                else if (!host.loadExtension) {
+                    error = new Error("Extension loading not implemented in host!");
+                }
+                if (error) {
+                    diagnostics.push(createCompilerDiagnostic(Diagnostics.Extension_loading_failed_with_error_0, error));
+                }
+                return { name, result, error };
+            });
+            const successfulExtensionLoadResults = filter(extensionLoadResults, res => !res.error);
+            const preparedExtensionObjects = map(successfulExtensionLoadResults, res => {
+                if (res.result) {
+                    return reduceProperties(res.result, (aggregate: Extension[], potentialExtension: BaseProviderStatic, key: string) => {
+                        if (!potentialExtension) {
+                            return; // Avoid errors on explicitly exported null/undefined (why would someone do that, though?)
+                        }
+                        const annotatedKind = potentialExtension["extension-kind"];
+                        if (typeof annotatedKind === "string") {

```

Comment:
>annotatedKind [](start = 35, length = 13)

when is this guard ever false? I thought "extension-kind" had a string literal union type.